### PR TITLE
tool: add repair command to ceph-dedup-tool

### PR DESCRIPTION
**Pull request by myoungwon**
Until now, ceph-dedup-tool only provides chunk-scrub
,which reports how many dangling references the object has,
but there is no cli command to fix the reference mismach.

Therefore, this commit adds repair command to correct detected
reference mismatch.

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/tools/ceph_dedup_tool.cc [hunk 1]
```diff
@@ -801,17 +812,79 @@ int chunk_scrub_common(const std::map < std::string, std::string > &opts,
...
       cls_cas_chunk_put_ref(op, oid);
+      ret = run_op(op, oid, object_name, chunk_io_ctx);
+    } else if (op_name == "chunk-repair") {
+      ret = rados.ioctx_create2(pool_id, io_ctx);
+      if (ret < 0) {
+	cerr << oid << " ref " << pool_id
+	     << ": referencing pool does not exist" << std::endl;
+	return ret;
+      }
+      int chunk_ref, base_ref;
```
Comment (athanatos) [inline src/tools/ceph_dedup_tool.cc:838]: Can we define these where they are initialized?

File: src/tools/ceph_dedup_tool.cc [hunk 2]
```diff
@@ -801,17 +812,79 @@ int chunk_scrub_common(const std::map < std::string, std::string > &opts,
...
+	if (base_ref == -ENOENT || base_ref == -ENOLINK) {
+	  base_ref = 0;
+	} else {
+	  return base_ref;
+	}
+      }
+      cout << target_object_name << " has " << base_ref << " reference for "
+	   << object_name << std::endl;
+      if (chunk_ref != base_ref) {
+	ceph_assert(base_ref < chunk_ref);
```
Comment (athanatos) [inline src/tools/ceph_dedup_tool.cc:874]: Let's print an error message and return an error code here.

**Comment by myoungwon** [comment]
@athanatos I addressed your comments.

File: src/tools/ceph_dedup_tool.cc [hunk 3]
```diff
@@ -801,17 +812,83 @@ int chunk_scrub_common(const std::map < std::string, std::string > &opts,
...
+      ret = rados.ioctx_create2(pool_id, io_ctx);
+      if (ret < 0) {
+	cerr << oid << " ref " << pool_id
+	     << ": referencing pool does not exist" << std::endl;
+	return ret;
+      }
+      int chunk_ref = -1, base_ref = -1;
+      // read object on chunk pool to know how many reference the object has
+      bufferlist t;
+      ret = chunk_io_ctx.getxattr(object_name, CHUNK_REFCOUNT_ATTR, t);
```
Comment (athanatos) [inline src/tools/ceph_dedup_tool.cc:841]: Are we supposed to go through a class operation here?

**Review by athanatos — APPROVED** [review]
LGTM other than my comment

File: src/tools/ceph_dedup_tool.cc (hunk 3 above)
Comment (athanatos) [reply src/tools/ceph_dedup_tool.cc:841]: Ok, seems reasonable.

**Comment by ronen-fr** [comment]
@myoungwon : can you take a look at the failure of
http://pulpito.front.sepia.ceph.com/yuriw-2021-11-03_15:48:02-rados-wip-yuri7-testing-2021-11-01-1748-distro-basic-smithi/6481640/
?
A similar test without this patch was OK. Might be related.

**Comment by yuriw** [comment]
@myoungwon this was tested and can be merged after further clarifying comments from above

**Comment by myoungwon** [comment]
@ronen-fr It seems that this is not related the failure as discussed at (https://github.com/ceph/ceph/pull/43411)
@neha-ojha @athanatos This is probably ready to merge, thoughts?

**Comment by athanatos** [comment]
@myoungwon I'm not sure what you mean.  The failure @yuriw pasted in is specific to this tool.  I think the following is the failure portion:

2021-11-07T06:11:56.720 INFO:tasks.workunit.client.0.smithi090.stderr:+ '[' -n 0 ']'
2021-11-07T06:11:56.720 INFO:tasks.workunit.client.0.smithi090.stderr:+ /usr/bin/ceph osd pool delete dedup_pool dedup_pool --yes-i-really-really-mean-it
2021-11-07T06:11:57.989 INFO:tasks.workunit.client.0.smithi090.stderr:pool 'dedup_pool' does not exist
2021-11-07T06:11:57.998 INFO:tasks.workunit.client.0.smithi090.stderr:+ /usr/bin/ceph osd pool delete dedup_chunk_pool dedup_chunk_pool --yes-i-really-really-mean-it
2021-11-07T06:11:58.994 INFO:tasks.workunit.client.0.smithi090.stderr:pool 'dedup_chunk_pool' does not exist
2021-11-07T06:11:59.004 INFO:tasks.workunit.client.0.smithi090.stderr:+ die 'Scrub failed expecting bar is removed'
2021-11-07T06:11:59.004 INFO:tasks.workunit.client.0.smithi090.stdout:Scrub failed expecting bar is removed
2021-11-07T06:11:59.005 DEBUG:teuthology.orchestra.run:got remote process result: 1
2021-11-07T06:11:59.005 INFO:tasks.workunit.client.0.smithi090.stderr:+ echo 'Scrub failed expecting bar is removed'
2021-11-07T06:11:59.006 INFO:tasks.workunit.client.0.smithi090.stderr:+ exit 1
2021-11-07T06:11:59.006 INFO:tasks.workunit:Stopping ['rados/test_dedup_tool.sh'] on client.0...

**Comment by myoungwon** [comment]
@athanatos My bad. I'll take a look.

**Comment by ljflores** [comment]
jenkins test make check

**Comment by myoungwon** [comment]
@athanatos I fixed test cases that caused the failure.

**Review by ljflores — COMMENTED** [review]
After the fixup changes are confirmed to work, I think the commits should then be squashed into one.

File: qa/workunits/rados/test_dedup_tool.sh [hunk 4]
```diff
@@ -48,16 +48,25 @@ OBJ=test_rados_obj
...
+  done	
 
   $RADOS_TOOL -p $POOL put $OBJ ./dedup_object_100k
-  RESULT=$($DEDUP_TOOL --op estimate --pool $POOL --chunk-size 1024  --chunk-algorithm fixed --fingerprint-algorithm sha1 --debug | grep result | awk '{print$4}')
-  if [ 1024 -ne $RESULT ];
+  RESULT=$($DEDUP_TOOL --op estimate --pool $POOL --chunk-size 1024  --chunk-algorithm fixed --fingerprint-algorithm sha1 | grep chunk_size_average | awk '{print$2}' | sed "s/\,//g")
+  # total size / the number of deduped object = 100K / 1 
+  if [ 51200 -ne $RESULT ];
   then
     die "Estimate failed expecting 1024 result $RESULT"
```
Comment (athanatos) [inline qa/workunits/rados/test_dedup_tool.sh:71]: 1024 wrong here?

Comment (athanatos) [reply qa/workunits/rados/test_dedup_tool.sh:71]: Also other messages.

**Review by athanatos — CHANGES_REQUESTED** [review]

**Comment by myoungwon** [comment]
Fixed and squashed.

**Review by athanatos — APPROVED** [review]

**Comment by athanatos** [comment]
Needs to be retested after the above changes.

**Comment by amathuria** [comment]
http://pulpito.front.sepia.ceph.com/yuriw-2021-11-20_18:01:41-rados-wip-yuri8-testing-2021-11-20-0807-distro-basic-smithi

Failures unrelated tracked in: 
https://tracker.ceph.com/issues/53287
https://tracker.ceph.com/issues/53345
https://tracker.ceph.com/issues/53353
https://tracker.ceph.com/issues/52279
https://tracker.ceph.com/issues/39150
https://tracker.ceph.com/issues/53288

Details:
Bug #53287: test_standby (tasks.mgr.test_prometheus.TestPrometheus) fails
Bug #53345: Test failure: test_daemon_restart (tasks.cephadm_cases.test_cli.TestCephadmCLI)
Bug #53353: mgr/dashboard: orchestrator/03-inventory.e2e-spec.ts failure
Bug #52279: cephadm tests fail due to: error adding seccomp filter rule for syscall bdflush: requested action matches default action of filter
Bug #39150: mon: "FAILED ceph_assert(session_map.sessions.empty())"  
Bug #53288: Failed jobs hanging for 12 hours (Dead jobs)

**Comment by myoungwon** [comment]
It seems that this PR is unrelated with the failures. Can you double-check? @athanatos

**Comment by neha-ojha** [comment]
jenkins test api

**Comment by ljflores** [comment]
jenkins test api

**Pull request merged**